Proposals/20200511-HVPA-design.md

```diff
@@ -0,0 +1,235 @@
+   * CPU: minimum: 500m, maximum: 9000m
+   * Memory: minimum: 1G, maximum: 20G
+
+The minimum resource value for a bucket is such that maximum aggregated resource from previous bucket is same as this bucket's minimum aggregated resource.
+
+### Notes/Constraints/Caveats
+
+#### VPA tuning
+HVPA controller is generous while switching buckets with the assumption that VPA would be tuned to react quickly enough to scale vertically down after the bucket is switched if the usage mandates it. For detailed reason, refer [this section](####base-usage-of-each-replica). The upstream changes done via [VPA PR #2905](https://github.com/kubernetes/autoscaler/pull/2905) would help in tuning VPA accordingly. These changes should be available in VPA releases 0.8.0+.
+Note: The global VPA settings would also be applicable for stand alone VPAs deployed for other components in the system.
+
+#### Different recommendations from HPA and VPA
+In such a case, higher recommendations will be used to calculate the desired replica count. This replica count would be used to calculate desired replica size.
+
+#### Different replica count from total CPU and total memory recommendations
+In such a case, higher replica count would be taken as the desired count. Desired memory and CPU per replica would be calculated by dividing desired total memory and total CPU by the chosen replica count respectively.
+
+#### Pod with multiple containers
+In such a case, all the calculations would be done using the total CPU and total memory allocated to all the containers in the pod for which VPA container resource policy is set to "Auto". The total recommended CPU and memory per replica would then be divided in the same proportion as the recommendations by VPA for those containers.
+
+#### Pod with multiple containers with VPA update resource policy for some containers set to "Off"
+With configuration like this, VPA may not want to scale, however, HPA may want to scale because of load on those containers. In such cases, HVPA controller would scale only horizontally as per HPA recommendations, and will ignore the bucket strategy altogether.
+For example:
+If there are no new VPA recommendations, and HPA recommends to scale out. It would mean that the total resource usage of all the containers has crossed the threshold provided to HPA, however, VPA does not see any need to change the resource requirements. There could be following possible scenarios:
+1. Resource usages for VPA enalbed containers have increased but VPA does not see any need to scale up the replicas just yet
+1. Total resource usage has increased due to containers for which VPA is not enabled
+
+Since HVPA does not keep any history of resource usages of containers, it would not be able to figure out the reason for such a disparity. In these conditions, HVPA would scale only horizontally as per HPA recommendations, and will ignore the bucket strategy so that unnecessary rolling updates could be avoided.
```

Comment:
The explanation clarifies some edge cases but not others. For example, if VPA was enabled for some containers and their recommendations dominated the total for all the containers and we still end up scaling vertically (possibly also horizontally if buckets say so).

It is unclear when and how the following statement will become effective.

> In these conditions, HVPA would scale only horizontally as per HPA recommendations, and will ignore the bucket strategy so that unnecessary rolling updates could be avoided.

Reply:
This case is only considering the condition when VPA has **no new recommendations**, and HPA recommends scaling, as clearly mentioned in opening lines:

> VPA may not want to scale, however, HPA may want to scale because of load on those containers. In such cases...

This is NOT applicable when VPA has **new** recommendations.

If following happens:
> if VPA was enabled for some containers and their recommendations dominated the total for all the containers

We ALWAYS take the bucket strategy.
I hope this clarifies.

Reply:
This is just to avoid unnecessary rolling updates. Since VPA is tuned to quickly react to any overscaling (as mentioned under section `VPA tuning`), it is free to recommend any new scaling.
Let me know if this is not really desired, and we can just scale as we would according to bucket strategy.
But this would result in overscaling anyway for those containers because their requirements have not really changed (because VPA has not reacted yet).
wdyt?

Reply:
Also, if we just follow bucket strategy, we might end up in the same bucket with same number of replicas but with higher `resources` for containers for which VPA is enabled. The containers because of which HPA recommendations came in would never see their load decreasing because for them, the `requests` and `replicas` remain the same.

Reply:
Reworded in https://github.com/gardener/hvpa-controller/pull/71/commits/cb08483fe4f0a0e5e560b2e151bd9f4ab6fedcd7 to make it more clear